tools/querytee/instant_query_transform_test.go

```diff
@@ -0,0 +1,171 @@
+
+package querytee
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+func TestAddMissingTimeParam(t *testing.T) {
+	// Mock the current time.
+	originalTimeNow := timeNow
+	timeNow = func() time.Time { return time.Date(2024, 6, 10, 20, 30, 40, 0, time.UTC) }
+	t.Cleanup(func() { timeNow = originalTimeNow })
+
+	testCases := map[string]struct {
+		method    string
+		urlParams url.Values
+		form      url.Values
+
+		expectedURLParams url.Values
+		expectedForm      url.Values
+	}{
+		"GET with time in URL": {
+			method: http.MethodGet,
+			urlParams: url.Values{
+				"query":   []string{"sum(abc)"},
+				"timeout": []string{"60s"},
+				"time":    []string{"2024-06-19T01:23:45Z"},
+			},
+
+			expectedURLParams: url.Values{
+				"query":   []string{"sum(abc)"},
+				"timeout": []string{"60s"},
+				"time":    []string{"2024-06-19T01:23:45Z"},
+			},
+			expectedForm: url.Values{},
+		},
+		"GET without time in URL": {
+			method: http.MethodGet,
+			urlParams: url.Values{
+				"query":   []string{"sum(abc)"},
+				"timeout": []string{"60s"},
+			},
+
+			expectedURLParams: url.Values{
+				"query":   []string{"sum(abc)"},
+				"timeout": []string{"60s"},
+				"time":    []string{"2024-06-10T20:30:40Z"},
+			},
+			expectedForm: url.Values{},
+		},
+		"POST with time in body": {
+			method: http.MethodPost,
+			form: url.Values{
+				"query":   []string{"sum(abc)"},
+				"timeout": []string{"60s"},
+				"time":    []string{"2024-06-19T01:23:45Z"},
+			},
+
+			expectedURLParams: url.Values{},
+			expectedForm: url.Values{
+				"query":   []string{"sum(abc)"},
+				"timeout": []string{"60s"},
+				"time":    []string{"2024-06-19T01:23:45Z"},
+			},
+		},
+		"POST without time in body": {
+			method: http.MethodPost,
+			form: url.Values{
+				"query":   []string{"sum(abc)"},
+				"timeout": []string{"60s"},
+			},
+
+			expectedURLParams: url.Values{},
+			expectedForm: url.Values{
+				"query":   []string{"sum(abc)"},
+				"timeout": []string{"60s"},
+				"time":    []string{"2024-06-10T20:30:40Z"},
+			},
+		},
+		"POST with time in URL": {
+			method: http.MethodPost,
+			urlParams: url.Values{
+				"query":   []string{"sum(abc)"},
+				"timeout": []string{"60s"},
+				"time":    []string{"2024-06-19T01:23:45Z"},
+			},
+
+			expectedURLParams: url.Values{
+				"query":   []string{"sum(abc)"},
+				"timeout": []string{"60s"},
+				"time":    []string{"2024-06-19T01:23:45Z"},
+			},
+			expectedForm: url.Values{},
+		},
+		"POST without time in URL": {
+			method: http.MethodPost,
+			urlParams: url.Values{
+				"query":   []string{"sum(abc)"},
+				"timeout": []string{"60s"},
+			},
+
+			expectedURLParams: url.Values{
+				"query":   []string{"sum(abc)"},
+				"timeout": []string{"60s"},
+				"time":    []string{"2024-06-10T20:30:40Z"},
+			},
+			expectedForm: url.Values{},
```

Comment:
Do you mean `time` is present in both? Or some parameters are in the URL and some are in the body?

Reply:
I meant some parameters are the URL and some in the body. I guess it could also be worth ensuring that if it is set in both that they are passed through unchanged.

Reply:
Added in 527c416.

If `time` is set in both, then I think all bets are off, so I've left that out for now.

Reply:
If it's set in both I would expect the set values to be passed on. What happens from there I'm unsure, but the former is still something we could check here. Not necessary though imo.